### PR TITLE
Main mrhone branch

### PR DESCRIPTION
**Pull request by Twooey**

### CODE REVIEW

File: unimus-backup-exporter.env [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#Unimus server address. Format http://[address]:[Port]
+unimus_server="http://foo.bar:8085"
+#Unimus API key
+unimus_api_key="insert api key here"
+#Unimus backup type. Valid options are latest or historical
```
Comment (tomaskir) [inline unimus-backup-exporter.env:5]: rename `historical` to `all`

File: unimus-backup-exporter.env [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+backup_type="latest"
```
Comment (tomaskir) [inline unimus-backup-exporter.env:6]: introduce `export_type`, setable by user to `git` or `fs`, fail if not set properly

File: unimus-backup-exporter.env [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+git_username="foo"
+#Git email
+git_email="foo@bar.org"
+#Git server protocal. Valid Options are ssh, http, https
+git_server_protocal="ssh"
+#Git server address. Format [address]
+git_server_address="192.168.4.5"
+#Git Server Port
+git_port="22"
+#Git Repo Address. Standared is [User][Repo Name]
```
Comment (tomaskir) [inline unimus-backup-exporter.env:17]: `for GitHub or GitLab standard is [User]/[Repo Name], for self-hosted repo [Repo Name]`

File: unimus-backup-exporter.env [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+git_repo_name="User/unimus-backup-extracter"
+#Git Branch
+git_branch="master"
+#Git Password
+git_password="password"
```
Comment (tomaskir) [inline unimus-backup-exporter.env:22]: move under `username`

File: unimus-backup-exporter.sh [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+#This is a Unimus to Git API to export your backups to your Git Repo
+
+
+script_dir=$(cd "$(dirname "${BASH_SOURCE[0]}")" &>/dev/null && pwd -P)
+backup_dir=$script_dir/backups
+
+
+
+#Imports from env
+set -a # automatically export all variables
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:12]: move config loading into function, do validation after loading

File: unimus-backup-exporter.sh [hunk 6]
```diff
@@ -0,0 +1,212 @@
+#!/usr/bin/env bash
+
+#This is a Unimus to Git API to export your backups to your Git Repo
+
+
+script_dir=$(cd "$(dirname "${BASH_SOURCE[0]}")" &>/dev/null && pwd -P)
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:6]: cleanup, all code either in functions, or at the bottom before the `main` call

File: unimus-backup-exporter.sh [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+source unimus-backup-exporter.env
+set +a
+
+
+#HashTable for all devices
+declare -A devices
+log=$script_dir/unimus-backup-exporter.log
+
+
+function echoGreen() { echo -e "${green}$1${reset}"; }
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:22]: consistent formatting across script

File: unimus-backup-exporter.sh [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+	local address=${devices[$1]}
+	local date="$2"
+	if [[ $4 == "TEXT" ]]; then
+		local type="txt"
+	elif [[ $4 == "BINARY" ]]; then
+		local type ="bin"
+	fi
+
+	if ! [ -d "backups" ]; then
+		mkdir backups
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:51]: this can fail, catch and deal with errors

File: unimus-backup-exporter.sh [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+	fi
+
+	if ! [ -d "$backup_dir/$address - $1" ]; then
+		mkdir $backup_dir/$address\ \-\ $1
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:55]: `"${backup_dir}/${address} - $1"`

File: unimus-backup-exporter.sh [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+	fi
+	if ! [ -e $backup_dir/$address\ -\ $1/Backup\ $address\ $date\ $1.$type ]; then
+		base64 -d <<< $3 > $backup_dir/$address\ -\ $1/Backup\ $address\ $date\ $1.$type
+	fi
+}
+
+function get_allDevices (){
+	echoGreen "Getting Device Information" >> $log
+	for ((i=0; ; i+=1)); do
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:64]: `while true` would be better

File: unimus-backup-exporter.sh [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+
+function pushToGit(){
+	cd $backup_dir
+	if ! [ "$(git rev-parse --is-inside-work-tree 2>/dev/null)" ]; then
+		git init 
+		git add . 
+		git commit -m "Initial Commit"
+		case $git_server_protocal in 
+			ssh)
+			git remote add orgin ssh://git@$git_server_address/$git_repo_name
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:145]: should not use hard-coded username, use from configuration

Comment (tomaskir) [reply unimus-backup-exporter.sh:145]: use password if provided in config, document in config that password should be empty to use key auth

Comment (tomaskir) [reply unimus-backup-exporter.sh:145]: automatically accept SSH thumbprint on new connections

File: unimus-backup-exporter.sh [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+			;;
+			https)
+			git remote add orgin https://$git_username:$git_password@$git_server_address:$git_port/$git_repo_name 
+			;;
+			*)
+			echoGreen "Invalid setting for git_server_protocal" 
+			;;
+		esac
+
+		git push -u orgin master
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:158]: use branch from config file instead of `master`

File: unimus-backup-exporter.sh [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+	fi
+	cd $script_dir
+}
+
+function main(){
+
+	#Creating a log file
+	printf "Log File - " >> $log
+	date >> $log
+
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:173]: spacing consistency

File: unimus-backup-exporter.sh [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+		git commit -m $"date" 
+		git push 
+	fi
+	cd $script_dir
+}
+
+function main(){
+
+	#Creating a log file
+	printf "Log File - " >> $log
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:171]: `"Log File - $(date)"` - explicitly define format for `date`

File: unimus-backup-exporter.sh [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+			echoGreen "Not Pushing to Git"; 
+		else 
+			pushToGit
+		fi
+	else
+		if [[ -z $status ]]; then
+			echoRed "Unable to Connect to server"
+		else
+
+			echoRed "Server Status: $status "
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:206]: formatting

File: unimus-backup-exporter.sh [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+			;;
+			*)
+			echoRed "Invalid Setting for backup_type" >> $log
+			;;
+		esac
+
+		#If no server protocal is selected we will not push to git.
+		#Otherwise We push to Git
+		if [ -z "$git_server_protocal" ]; then 
+			echoGreen "Not Pushing to Git";
```
Comment (tomaskir) [inline unimus-backup-exporter.sh:197]: not needed, we don't want to spam the log for users not using GIT

File: README.md [hunk 17]
```diff
@@ -0,0 +1,35 @@
+# Unimus Backup Extractor
+
+This repository contains the Unimus Backup Extractor.  
+This extractor is used to create local copies of unimus device backups. 
+
+### TL;DR
+Requirements: `bash`, `jq`, `base64`.
+```text
```
Comment (tomaskir) [inline README.md:8]: should show user how to get functional - which requires 2 steps, config setup, running script

File: README.md [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+./unimus-backup-exporter.sh
+``` 
+
+### How does the exporter work?
```
Comment (tomaskir) [inline README.md:12]: in this section, describe how this works, uses Unimus API, exports backups either `latest` or `all`, etc.

File: README.md [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+
+The Following are valid options for the .env file. 
+
+    unimus_server=http://[address]:[Port]
+    unimus_api_key="api_key"
+    backup_type="[latest] or [historical]"
+
+Valid options for git settings are as followed. If you do not wish to push to git, you can remove these, or put "none" in the git_server_protocal
+
+    git_server_protocal="[http], [https], [ssh], or [none]"
```
Comment (tomaskir) [inline README.md:24]: Add a separate `Configuration file` section, describe everything related to config there

**Review by tomaskir — CHANGES_REQUESTED** [review]

File: unimus-backup-exporter.env (hunk 1 above)
Comment (Twooey) [reply unimus-backup-exporter.env:5]: historical has been renamed to all

File: unimus-backup-exporter.env (hunk 4 above)
Comment (Twooey) [reply unimus-backup-exporter.env:22]: Password has been moved to under username

File: unimus-backup-exporter.sh (hunk 5 above)
Comment (Twooey) [reply unimus-backup-exporter.sh:12]: Loading moved into function, Validation added.

File: unimus-backup-exporter.sh (hunk 7 above)
Comment (Twooey) [reply unimus-backup-exporter.sh:22]: Formatting has been made more consistant. 

Also not listed in this PR, echoGreen, echoYellow, and echoRed now display proper colors.

File: unimus-backup-exporter.sh (hunk 6 above)
Comment (Twooey) [reply unimus-backup-exporter.sh:6]: Code has been reorganized.

File: unimus-backup-exporter.sh (hunk 8 above)
Comment (Twooey) [reply unimus-backup-exporter.sh:51]: Errors are now caught and handled.

File: unimus-backup-exporter.sh (hunk 9 above)
Comment (Twooey) [reply unimus-backup-exporter.sh:55]: Partially fixed. Still working on another bug.

File: unimus-backup-exporter.sh (hunk 10 above)
Comment (Twooey) [reply unimus-backup-exporter.sh:64]: I was incorrect here. We use the for loop to traverse through pages and each json entry.

File: unimus-backup-exporter.sh (hunk 11 above)
Comment (Twooey) [reply unimus-backup-exporter.sh:145]: username is not hard coded. Still need to implement ssh password or key, depending on the situation. As well as an option to auto-accept ssh thumbprints.

File: unimus-backup-exporter.sh (hunk 12 above)
Comment (Twooey) [reply unimus-backup-exporter.sh:158]: Script now uses config file's branch variable instead of hardcoded master.

**Pull request merged**